Enable nesting for python_mode

samdow opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75965

This adds the ability to nest python_modes with `push_python_mode`, based on the torch_function design introduced [here](https://github.com/pytorch/pytorch/pull/75154). In order to push python modes, we needed to (1) introduce the concept of PythonMode (a la TorchFunctionMode) that defines a torch_dispatch but isn't a tensor subclass, (2) update enable_python_mode to match torch function mode (longer explanation below), and (3) introduce `push_python_mode` alongside the now updated `enable_python_mode`. Users will only be able to nest modes using push but not enable

**Longer note on what it means to update enable_python_mode**

`enable_python_mode`/`enable_torch_function_mode` are the simplest versions of the modes that don't support nesting but do support passing a tensor subclass as the mode. The current implementation of `enable_python_mode` didn't match the functionality of `enable_torch_function_mode` (which limited the ability to add nesting in the form of `push_python_mode` in the next PR)

This makes their behaviors match which means:
- the mode is disabled during the call to torch_dispatch, this removes the `no_dispatch` calls that needed to be everywhere in order to prevent infinite recursion and will be necessary for nested modes
- nested modes are "allowed" (doesn't break) if the mode passed to the instance is the same (it's treated as a noop in this case). Otherwise, it will error like before
- Adds some `kwargs` that let a user override a current mode to a new one (not the same as nesting because it ignores the outer context manager)
- small string differences so the error messages are the same between the two

Commit:
update enable_python_mode to match enable_torch_function_mode

[ghstack-poisoned]

samdow commented on torch/csrc/autograd/init.cpp:
```diff
@@ -588,32 +599,46 @@ static PyObject * get_torch_function_mode(PyObject* _unused, PyObject* _unused2)
 
 // autograd methods on torch._C
 static PyMethodDef methods[] = { // NOLINT
-  {"_set_grad_enabled", set_grad_enabled, METH_O, nullptr},
-  {"is_grad_enabled", is_grad_enabled, METH_NOARGS, nullptr},
-  {"is_inference_mode_enabled", is_inference_mode_enabled, METH_NOARGS, nullptr},
-  {"set_autocast_enabled", set_autocast_enabled, METH_O, nullptr},
-  {"is_autocast_enabled", is_autocast_enabled, METH_NOARGS, nullptr},
-  {"clear_autocast_cache", clear_autocast_cache, METH_NOARGS, nullptr},
-  {"set_autocast_cpu_enabled", set_autocast_cpu_enabled, METH_O, nullptr},
-  {"is_autocast_cpu_enabled", is_autocast_cpu_enabled, METH_NOARGS, nullptr},
-  {"set_autocast_cpu_dtype", set_autocast_cpu_dtype, METH_O, nullptr},
-  {"get_autocast_cpu_dtype", get_autocast_cpu_dtype, METH_NOARGS, nullptr},
-  {"set_autocast_gpu_dtype", set_autocast_gpu_dtype, METH_O, nullptr},
-  {"get_autocast_gpu_dtype", get_autocast_gpu_dtype, METH_NOARGS, nullptr},
-  {"autocast_increment_nesting", autocast_increment_nesting, METH_NOARGS, nullptr},
-  {"autocast_decrement_nesting", autocast_decrement_nesting, METH_NOARGS, nullptr},
-  {"is_autocast_cache_enabled", is_autocast_cache_enabled, METH_NOARGS, nullptr},
-  {"set_autocast_cache_enabled", set_autocast_cache_enabled, METH_O, nullptr},
-  {"set_anomaly_enabled", set_anomaly_mode_enabled, METH_O, nullptr},
-  {"is_anomaly_enabled", is_anomaly_mode_enabled, METH_NOARGS, nullptr},
-  {"_enter_dual_level", python_enter_dual_level, METH_NOARGS, nullptr},
-  {"_exit_dual_level", castPyCFunctionWithKeywords(python_exit_dual_level), METH_VARARGS | METH_KEYWORDS, nullptr},
-  {"_enter_python_mode", enter_python_mode, METH_O, nullptr},
-  {"_exit_python_mode", exit_python_mode, METH_NOARGS, nullptr},
-  {"_set_torch_function_mode", set_torch_function_mode, METH_O, nullptr},
-  {"_get_torch_function_mode", get_torch_function_mode, METH_NOARGS, nullptr},
-  {nullptr, nullptr, 0, nullptr}
-};
```
seems like clang-tidy had too much fun

revert

samdow commented on torch/mode_classes.py:
```diff
@@ -0,0 +1,170 @@
+import functools
+
+# Implementation note: I had a choice about how much of mode stacks
+# to implement in Python versus in C++.  At time of writing, I did not care
+# too much about implementation efficiency; however, I do care about making it
+# hard for users to implement modes in the wrong way.  In the end, it turned
+# out to be possible to implement mode stacks entirely from userland, with the
+# C++ API providing only _get_torch_function_mode() and
+# _set_torch_function_mode(), so I opted to provide some unsafe C++ bindings and
+# have the bulk of the logic for managing the stack in Python, which helped
+# simplify the C++ API surface.  It would also have been valid to build in the
+# notion of mode stack directly into C++ but in this design it's substantially
+# more difficult to interact with TorchFunctionModeMeta.
```
I think this is in the modes themselves which even though it's duplicated, it should stay there

remove

samdow commented on torch/mode_classes.py:
```diff
@@ -0,0 +1,170 @@
+        return super().__new__(metacls, name, bases, dct)
+
+
+class PythonMode(metaclass=PythonModeMeta):
```
remove. These should be added by the next PR because their main use is for nested modes

samdow commented on torch/utils/_mode_utils.py:
```diff
@@ -0,0 +1,87 @@
+from torch._C import (
+    _get_torch_function_mode, _set_torch_function_mode, _get_python_mode, _set_python_mode)
+
+def _wrap_init(f, metaclass, mode_type):
+    undef = object()
+
+    @functools.wraps(f)
+    def wrapped(self, *args, inner=undef, **kwargs):
+        if inner is undef:
+            assert metaclass.__name__[-4:] == "Meta"
+            non_meta_class = metaclass.__name__[:-4]
+            raise TypeError(
+                f"missing inner keyword argument; instead of constructing a {non_meta_class} directly, "
+                f"pass the constructor to push_{mode_type}_mode"
+            )
+        self.inner = inner
+        return f(self, *args, **kwargs)
+    return wrapped
+
+class ModeMeta(type):
+    def __new__(metacls, name, bases, dct, mode_type=None):
+        if mode_type not in ['torch_function', 'python']:
+            raise RuntimeError(f"only support torch_function or python modes, got mode_type of {mode_type}")
+        if '__init__' in dct:
+            dct['__init__'] = _wrap_init(dct['__init__'], metacls, mode_type)
+        return super().__new__(metacls, name, bases, dct)
```
remove--should be added in next PR

Commit:
Update on "update enable_python_mode to match enable_torch_function_mode"




[ghstack-poisoned]

Commit:
Update on "update enable_python_mode to match enable_torch_function_mode"




[ghstack-poisoned]

Commit:
Update on "update enable_python_mode to match enable_torch_function_mode"




[ghstack-poisoned]

Commit:
Update on "update enable_python_mode to match enable_torch_function_mode"




[ghstack-poisoned]

zou3519 commented on aten/src/ATen/core/PythonFallbackKernel.cpp:
```diff
@@ -62,9 +62,13 @@ void pythonFallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
 
   // If Python Mode is active, use its PyInterpreter for dispatch
-  const auto& maybe_python_mode_state = at::impl::PythonModeTLS::get_state();
+  // need a copy, not a reference because we will set the state to none before
```
What exactly do you mean by "update enable_python_mode to match enable_torch_function_mode"? Some more context in the PR body would help

zou3519 commented:
cc @ezyang if you wanted to take a look as well

zou3519 commented on torch/csrc/autograd/init.cpp:
```diff
@@ -616,8 +627,8 @@ static PyMethodDef methods[] = { // NOLINT
   {"_exit_dual_level", castPyCFunctionWithKeywords(python_exit_dual_level), METH_VARARGS | METH_KEYWORDS, nullptr},
-  {"_enter_python_mode", enter_python_mode, METH_O, nullptr},
-  {"_exit_python_mode", exit_python_mode, METH_NOARGS, nullptr},
+  {"_set_python_mode", set_python_mode, METH_O, nullptr},
+  {"_get_python_mode", get_python_mode, METH_NOARGS, nullptr},
```
nit: we probably want to start referring to these as set_torch_dispatch_mode / get_torch_dispatch_mode to differentiate it from torch_function mode. "python mode" by itself is ambiguous. Doesn't need to happen in this PR

zou3519 commented on torch/_C/__init__.pyi.in:
```diff
@@ -743,8 +743,10 @@ def __set_forward_AD_enabled(enabled: _bool) -> None: ...
 def _reset_default_hooks() -> None: ...
-def _enter_python_mode(cls: Type) -> None: ...
-def _exit_python_mode() -> None: ...
+def _set_torch_function_mode(cls: Type) -> None: ...
```
Should cls be union[TorchFunctionMode, Type] ? Also I am not sure if you applied the same treatment for torch_dispatch mode (still going through this)

samdow replied:
Oh good point...it errored in the second PR probably because it wasn't that union. I wonder if we'll end up with a circular dependency if this needs the Mode types? Well, let's see!

zou3519 commented on aten/src/ATen/core/PythonFallbackKernel.cpp:
```diff
@@ -62,9 +62,13 @@ void pythonFallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
+  // dispatching
+  const auto maybe_python_mode_state = at::impl::PythonModeTLS::get_state();
   if (maybe_python_mode_state) {
+    at::impl::PythonModeTLS::set_state(nullptr);
     maybe_python_mode_state->pyinterpreter()->dispatch(op, stack, maybe_python_mode_state);
+    at::impl::PythonModeTLS::set_state(maybe_python_mode_state);
```
To summarize... when we get to the __torch_dispatch__ , the mode is disabled?

1) Is there a way to turn the mode back on in torch_dispatch? (or do we not care about that at all?)
2) Does this need a RAII guard? If dispatch raises an error, then the mode is going to be off completely. (If so, this could be a test case)

samdow replied:
> To summarize... when we get to the torch_dispatch , the mode is disabled?

Yes--this part matches the torch_function version

> Is there a way to turn the mode back on in torch_dispatch? (or do we not care about that at all?)

Yes--this is actually similar to what will happen for nesting. In that case, the outer mode's torch_dispatch will end up getting wrapped in a mode call so it will be:

outer mode hits this call -> disable mode -> outer mode's torch dispatch wrapper -> outer mode enables inner mode (we can do this because the mode's disabled) -> call outer's torch_dispatch with inner mode enabled

> Does this need a RAII guard? If dispatch raises an error, then the mode is going to be off completely. (If so, this could be a test case)

Hmmm good point, I think so yes. There are already a lot of tests that raise errors though so I'm a little interested that we haven't already hit that

albanD replied:
You can copy paste some of the RAII above in this file to do that easily.

samdow replied:
RAII added

albanD replied:
Looks good

zou3519 commented on test/test_python_dispatch.py:
```diff
@@ -517,11 +514,82 @@ def test_enable_python_mode_respects_no_dispatch(self) -> None:
+                with enable_python_mode(A):
                     pass
 
+        # nested enable_python_modes are allowed if they're the same mode
+        # Will only write once to the log
+        with capture_logs() as logs:
+            x = LoggingTensor(torch.tensor([3.]))
+            log_input("x", x)
+            with enable_python_mode(LoggingTensor):
+                with enable_python_mode(LoggingTensor):
+                    x + x
```
Does torch_function mode have the same behavior?

I'm not sure how I feel about this. It's clearer if both things passed in to enable_python_mode are actually Mode objects, because then if two different mode instances are passed in then that counts as nesting mode but if it is the same mode instance then this behavior seems intuitive

samdow replied:
Yes. I'll split this test function so it's clearer. Nesting does _not_ work with `enable_python_mode` generally (this is tested in the lines above), but there's a special case where it will ignore the second context manager if the mode/tensor subclass passed to it is the same

zou3519 commented on torch/utils/_python_dispatch.py:
```diff
@@ -1,34 +1,51 @@
+def enable_python_mode(mode, *, replace=None, ignore_preexisting=False) -> Iterator[None]:
+    """
+    Context manager that causes all pytorch operators to dispatch to the passed-in
+    type's __torch_dispatch__ function, including operations that accepts no tensors
```
nit: "operations that accept"

zou3519 commented on torch/utils/_python_dispatch.py:
```diff
@@ -1,34 +1,51 @@
+    APIs recursively call back into your mode handler (this can easily cause
+    infinite loops, so use with care!)
+
+    enable_python_mode is affected by _DisableTorchDispatch.
```
Lol _DisableTorchDispatch is not public API, but I guess that's OK, we haven't released enable_python_mode as public API yet either

ezyang replied:
BTW, this is not symmetric with how DisableTorchFunction interacts with torch function mode. Disabling torch function means we no longer try to dispatch off of arguments, but a mode may still be valid and this situation could occur if the mode had originally returned NotImplemented for tensor subclass arguments

samdow replied:
> BTW, this is not symmetric with how DisableTorchFunction interacts with torch function mode. Disabling torch function means we no longer try to dispatch off of arguments, but a mode may still be valid and this situation could occur if the mode had originally returned NotImplemented for tensor subclass arguments

Not sure if we want to unify on one behavior, but I think had some cases in AOTAutograd tracing where it was important that the mode also disabled the mode behavior. I'll try to isolate the cases to better explain

ezyang replied:
Will be very interested to see it. My default assumption is that AOTAutograd was relying on the behavior in a bad way and writing the implementation slightly different would resolve it. If you can't isolate it I'm also happy to look into the complicated situation.

zou3519 commented on torch/utils/_python_dispatch.py:
```diff
@@ -1,34 +1,51 @@
+    enable_python_mode is affected by _DisableTorchDispatch.
+
+    Args:
+        mode (:class:`PythonMode`, Tensor-like class or None): the
```
Hold up, does this PR allow torch_dispatch mode to accept a Mode object?

zou3519 replied:
Doesn't look like it

samdow replied:
Whoops--sorry forgot to change this comment as I was splitting the commits. Thanks for the flag!

zou3519 commented on torch/utils/_mode_utils.py:
```diff
@@ -0,0 +1,62 @@
+
+# a class for the helper function to package all the info about the Mode class
+# so the helper function can access it without needing a circular import
+class _ModeInfo:
+    def __init__(self, mode_type: str, mode_class: type, base_mode_class: type, mode_class_name: str):
+        if mode_type not in ['torch_function', 'python']:
+            raise RuntimeError(f"only support torch_function or python modes, got mode_type of {mode_type}")
+        self.is_torch_function_mode = mode_type == 'torch_function'
+        self.mode_type = mode_type
+        self.mode_class = mode_class
+        self.base_mode_class = base_mode_class
+        self.mode_class_name = mode_class_name
+
+def _enable_mode(mode, mode_info: _ModeInfo, *, replace=None, ignore_preexisting=False) -> Iterator[None]:
+    if mode_info.is_torch_function_mode:
+        allowed_types = (
+            isinstance(mode, mode_info.mode_class) or
+            (isinstance(mode, type) and not issubclass(mode, mode_info.mode_class)))
+    else:
+        allowed_types = isinstance(mode, type) and issubclass(mode, (torch.Tensor,))
+    if not (mode is None or allowed_types):
+        if mode_info.is_torch_function_mode:
+            allowed_options = "TorchFunctionMode, Tensor-like class, or None"
+        else:
+            allowed_options = "Tensor-like class or None"
+        raise ValueError(f'expected to get {allowed_options} as argument, got {type(mode)} instead')
+    old = _get_torch_function_mode() if mode_info.is_torch_function_mode else _get_python_mode()
+    if old is mode:
```
I don't know if this is overkill (trying to merge the implementations of enable_torch_function_mode and enable_torch_dispatch_mode). We're not sure what the long-term interaction of both of these things together will look like.

If you do want to go down this route, having all of these if-then-else-blah is not the cleanest thing in the world. A better idiom might be to have:
- ModeInfo class that has more extension points (e.g., it can have a "supports_input(input) -> bool" method
- TorchFunctionModeInfo class that subclasses ModeInfo with all of the overrides
- TorchDispatchModeInfo class that subclasses ModeInfo
- _enable_mode implementation accepts a ModeInfo class and only queries attributes or calls methods on ModeInfo. It shouldn't need to if/then/else switching on whether the ModeInfo comes from torchfunction or torchdispatch

samdow replied:
I think the stronger argument for having them merged is for the nested version (`push_{torch_function|python}_mode`). Because we're currently doing the mode determination in python, it would be great for any performance improvements or moves to C++ to benefit both types of modes. It seems likely that if I were writing an update, I would forget to update both of them

Having the merged implementations for this was kind of a product of merging the nested modes and then not wanting to have the unevenness of one of them being merged and one of them not. But open if it still feels like overkill

Updated ModeInfo though!

ezyang replied:
I'm OK with linking up the designs here.

zou3519 replied:
+1, the deduplication sounds good and we can always un-deduplicate if it comes to that in the future

zou3519 reviewed (commented):
Seems reasonable, I had some questions and comments.

As a meta-point: I didn't know what "update enable_python_mode to match enable_torch_function_mode" meant before I read through the PR but in retrospect it makes sense now 🤦. Maybe a brief explanation in the PR body would be good?

Commit:
Update on "update enable_python_mode to match enable_torch_function_mode"


`enable_python_mode`/`enable_torch_function_mode` (torch_function_mode introduced [here](https://github.com/pytorch/pytorch/pull/75154)) are the simplest versions of the modes that don't support nesting but do support passing a tensor subclass as the mode. The current implementation of `enable_python_mode` didn't match the functionality of `enable_torch_function_mode` (which limited the ability to add nesting in the form of `push_python_mode` in the next PR)

This makes their behaviors match which means:
- the mode is disabled during the call to torch_dispatch, this removes the `no_dispatch` calls that needed to be everywhere in order to prevent infinite recursion and will be necessary for nested modes
- nested modes are "allowed" (doesn't break) if the mode passed to the instance is the same (it's treated as a noop in this case). Otherwise, it will error like before
- Adds some `kwargs` that let a user override a current mode to a new one (not the same as nesting because it ignores the outer context manager)
- small string differences so the error messages are the same between the two

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/init.cpp:
```diff
@@ -555,17 +554,29 @@ static PyObject * python_exit_dual_level(PyObject* _unused, PyObject* args, PyOb
 }
 
-static PyObject * enter_python_mode(PyObject* _unused, PyObject* arg) {
+static PyObject* set_python_mode(PyObject* _unused, PyObject* arg) {
```
Should we name these `set_torch_dispatch_mode` to improve consistency?

albanD replied:
Yes, different PR sounds good

albanD commented on test/test_python_dispatch.py:
```diff
@@ -517,11 +514,83 @@ def test_enable_python_mode_respects_no_dispatch(self) -> None:
 
     def test_nested_enable_python_mode(self) -> None:
-        with self.assertRaisesRegex(RuntimeError, "has already been set"):
+        class ErrorA(RuntimeError):
```
All of this is not really needed since we just check that it fails to enable the mode? We could simply subclass LoggingTensor here.

albanD commented on test/test_python_dispatch.py:
```diff
@@ -517,11 +514,83 @@ def test_enable_python_mode_respects_no_dispatch(self) -> None:
+            def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+                return Subclass_41(torch.zeros(()))
+
+        class Subclass_40(torch.Tensor):
```
Nit:
```
class Subclass_40(Subclass_41):
  pass
```

ezyang commented on torch/utils/_mode_utils.py:
```diff
@@ -0,0 +1,54 @@
+
+# a class for the helper function to package all the info about the Mode class
+# so the helper function can access it without needing a circular import
+class _ModeInfo:
```
Is there a design for ModeInfo anywhere? I suppose I'll find out more about what it's used for as I read this PR, but the comment here doesn't really help (and we didn't have anything like ModeInfo for torch function)

ezyang replied:
Suggestion: make this a dataclass

samdow replied:
done!

ezyang commented on torch/utils/_mode_utils.py:
```diff
@@ -0,0 +1,54 @@
+            f'Attempted to enable_{mode_info.mode_type}_mode, but there is already an '
+            f'active mode {old}.  {mode_info.help_text(mode)}'
+        )
+    # NB: we don't require TorchFunctionMode/PythonMode since this is intended to also
```
~~TorchFunctionMode~~ for this logic, right?

ezyang replied:
OHH I see what you did, you deduplicated the logic

samdow replied:
More discussion [here](https://github.com/pytorch/pytorch/pull/75965#discussion_r857708960) on if it's worth it to dedupe. I'll add more comments to ModeInfo and _enable_mode so it's easier to understand

ezyang commented on torch/utils/_mode_utils.py:
```diff
@@ -0,0 +1,54 @@
+        raise NotImplementedError()
+
+    def set_mode(self, mode):
+        raise NotImplementedError()
```
Highly recommend having docs for all of the abstract methods here

Commit:
Update on "update enable_python_mode to match enable_torch_function_mode"


`enable_python_mode`/`enable_torch_function_mode` (torch_function_mode introduced [here](https://github.com/pytorch/pytorch/pull/75154)) are the simplest versions of the modes that don't support nesting but do support passing a tensor subclass as the mode. The current implementation of `enable_python_mode` didn't match the functionality of `enable_torch_function_mode` (which limited the ability to add nesting in the form of `push_python_mode` in the next PR)

This makes their behaviors match which means:
- the mode is disabled during the call to torch_dispatch, this removes the `no_dispatch` calls that needed to be everywhere in order to prevent infinite recursion and will be necessary for nested modes
- nested modes are "allowed" (doesn't break) if the mode passed to the instance is the same (it's treated as a noop in this case). Otherwise, it will error like before
- Adds some `kwargs` that let a user override a current mode to a new one (not the same as nesting because it ignores the outer context manager)
- small string differences so the error messages are the same between the two

[ghstack-poisoned]

ezyang commented on torch/overrides.py:
```diff
@@ -1785,6 +1787,37 @@ def _no_torch_function_mode() -> Iterator[None]:
+                         required_fn="__torch_function__")
+
+    def is_allowed_type(self, mode) -> bool:
+        return mode is None or\
```
wow, I'm surprised the linter didn't complain about backslash after keyword with no intervening space lol

ezyang commented on torch/utils/_mode_utils.py:
```diff
@@ -0,0 +1,69 @@
+    mode_name: str
+    mode_class: type  # the class related to the mode that's allowed to be passed in
+    base_mode_class: type  # the base class of mode_class that dispatches to the original function
+    required_fn: str  # string version of the function required, either torch_function or torch_dispatch
```
Actually, I'm confused why these are fields. These fields (or at least `mode_class`) don't appear to be part of the public API of `_ModeInfo`, and they're only used in the downstream abstract method implementations. If they don't need to be in the public API, it seems clearer and more straightforward for the downstream subclass to take care of organizing things as necessary. As it stands, these fields seem quite unmotivated.

ezyang commented on torch/utils/_python_dispatch.py:
```diff
@@ -2,33 +2,76 @@
+                         required_fn="__torch_dispatch__")
+
+    def is_allowed_type(self, mode) -> bool:
+        return mode is None or isinstance(mode, type) and issubclass(mode, (torch.Tensor,))
```
You're not going to require torch dispatch modes to inherit from a TorchDispatchMode superclass? (If you did, then it would have been possible to shared `is_allowed_type` across the two implementations.

ezyang commented on torch/overrides.py:
```diff
@@ -1785,6 +1787,37 @@ def _no_torch_function_mode() -> Iterator[None]:
+    def is_allowed_type(self, mode) -> bool:
+        return mode is None or\
+            isinstance(mode, self.mode_class) or\
+            (isinstance(mode, type) and not issubclass(mode, self.mode_class))
```
What if I changed my mind and said I don't want direct Tensor-like modes anymore. Then it would be possible to share this code exactly between the two variants (with only `mode_class` varying).

ezyang replied:
and then the allowed types error message would be simpler

ezyang commented on torch/overrides.py:
```diff
@@ -1785,6 +1787,37 @@ def _no_torch_function_mode() -> Iterator[None]:
+                'pass ignore_preexisting=True.  This can result in unexpected '
+                'behavior; please consider rewriting your mode to be a subclass '
+                f'of {self.mode_class.__name__} to make it compositional!'
+            )
```
This error text feels like it should be shared between the two modes, for the most type

ezyang commented:
I feel like there is a big high level problem with this PR: where's the mode stack! (Maybe you were planning to add this on a later stack? Might be easier to just do it all in one go; I think it will be informative for some of the abstractions you're working on.)

samdow commented:
> I feel like there is a big high level problem with this PR: where's the mode stack! (Maybe you were planning to add this on a later stack?
Yep. That's added in #75966

> Might be easier to just do it all in one go; I think it will be informative for some of the abstractions you're working on.)

Okay--happy to squash the two PRs. I was worried it was going to be too big to do in one go but if it seems easier that's fine

Commit:
Update on "update enable_python_mode to match enable_torch_function_mode"


`enable_python_mode`/`enable_torch_function_mode` (torch_function_mode introduced [here](https://github.com/pytorch/pytorch/pull/75154)) are the simplest versions of the modes that don't support nesting but do support passing a tensor subclass as the mode. The current implementation of `enable_python_mode` didn't match the functionality of `enable_torch_function_mode` (which limited the ability to add nesting in the form of `push_python_mode` in the next PR)

This makes their behaviors match which means:
- the mode is disabled during the call to torch_dispatch, this removes the `no_dispatch` calls that needed to be everywhere in order to prevent infinite recursion and will be necessary for nested modes
- nested modes are "allowed" (doesn't break) if the mode passed to the instance is the same (it's treated as a noop in this case). Otherwise, it will error like before
- Adds some `kwargs` that let a user override a current mode to a new one (not the same as nesting because it ignores the outer context manager)
- small string differences so the error messages are the same between the two

[ghstack-poisoned]

samdow commented:
@ezyang Squashed the PRs--most of the error messages are able to be shared. Let me know what you think!

cc @albanD @zou3519 if either of you want to take a look again. PR comment is updated to reflect the new changes

Commit:
Update on "Enable nesting for python_mode"


This adds the ability to nest python_modes with `push_python_mode`, based on the torch_function design introduced [here](https://github.com/pytorch/pytorch/pull/75154). In order to push python modes, we needed to (1) introduce the concept of PythonMode (a la TorchFunctionMode), (2) update enable_python_mode to match torch function mode (longer explanation below), and (3) introduce `push_python_mode` alongside the now updated `enable_python_mode`. Users will only be able to nest modes using push but not enable.

**Longer note on what it means to update enable_python_mode**

`enable_python_mode`/`enable_torch_function_mode` are the simplest versions of the modes that don't support nesting but do support passing a tensor subclass as the mode. The current implementation of `enable_python_mode` didn't match the functionality of `enable_torch_function_mode` (which limited the ability to add nesting in the form of `push_python_mode` in the next PR)

This makes their behaviors match which means:
- the mode is disabled during the call to torch_dispatch, this removes the `no_dispatch` calls that needed to be everywhere in order to prevent infinite recursion and will be necessary for nested modes
- nested modes are "allowed" (doesn't break) if the mode passed to the instance is the same (it's treated as a noop in this case). Otherwise, it will error like before
- Adds some `kwargs` that let a user override a current mode to a new one (not the same as nesting because it ignores the outer context manager)
- small string differences so the error messages are the same between the two

[ghstack-poisoned]

albanD commented on aten/src/ATen/core/PythonFallbackKernel.cpp:
```diff
@@ -56,14 +56,32 @@ struct StashTLSOnEntryGuard {
+  const auto maybe_python_mode_state = at::impl::PythonModeTLS::get_state();
   if (maybe_python_mode_state) {
+    StashPythonModeGuard pythonModeGuard;
+    std::cout << "HERE" << std::endl;
```
Here?

samdow replied:
whoops! removed thanks!

albanD reviewed (commented):
At a high level, I really think we should rename `PythonMode` into `DispatchMode`. And while I agree that the private APIs can be renamed later. If we're adding public ones here, we should have them with the right name to begin with.

Commit:
Update on "Enable nesting for python_mode"


This adds the ability to nest python_modes with `push_python_mode`, based on the torch_function design introduced [here](https://github.com/pytorch/pytorch/pull/75154). In order to push python modes, we needed to (1) introduce the concept of PythonMode (a la TorchFunctionMode) that defines a torch_dispatch but isn't a tensor subclass, (2) update enable_python_mode to match torch function mode (longer explanation below), and (3) introduce `push_python_mode` alongside the now updated `enable_python_mode`. Users will only be able to nest modes using push but not enable

**Longer note on what it means to update enable_python_mode**

`enable_python_mode`/`enable_torch_function_mode` are the simplest versions of the modes that don't support nesting but do support passing a tensor subclass as the mode. The current implementation of `enable_python_mode` didn't match the functionality of `enable_torch_function_mode` (which limited the ability to add nesting in the form of `push_python_mode` in the next PR)

This makes their behaviors match which means:
- the mode is disabled during the call to torch_dispatch, this removes the `no_dispatch` calls that needed to be everywhere in order to prevent infinite recursion and will be necessary for nested modes
- nested modes are "allowed" (doesn't break) if the mode passed to the instance is the same (it's treated as a noop in this case). Otherwise, it will error like before
- Adds some `kwargs` that let a user override a current mode to a new one (not the same as nesting because it ignores the outer context manager)
- small string differences so the error messages are the same between the two

[ghstack-poisoned]

samdow commented:
> At a high level, I really think we should rename PythonMode into DispatchMode. And while I agree that the private APIs can be renamed later. If we're adding public ones here, we should have them with the right name to begin with.

IIUC, this entire API is actually prototype/private right now. If we were to do this, we should also change `enable_python_mode` and `push_python_mode` just because those will also end up being user facing. I'll try to look at a change before Ed and Richard look at this but just wanted to call this out

Commit:
Update on "Enable nesting for python_mode"


This adds the ability to nest python_modes with `push_python_mode`, based on the torch_function design introduced [here](https://github.com/pytorch/pytorch/pull/75154). In order to push python modes, we needed to (1) introduce the concept of PythonMode (a la TorchFunctionMode) that defines a torch_dispatch but isn't a tensor subclass, (2) update enable_python_mode to match torch function mode (longer explanation below), and (3) introduce `push_python_mode` alongside the now updated `enable_python_mode`. Users will only be able to nest modes using push but not enable

**Longer note on what it means to update enable_python_mode**

`enable_python_mode`/`enable_torch_function_mode` are the simplest versions of the modes that don't support nesting but do support passing a tensor subclass as the mode. The current implementation of `enable_python_mode` didn't match the functionality of `enable_torch_function_mode` (which limited the ability to add nesting in the form of `push_python_mode` in the next PR)

This makes their behaviors match which means:
- the mode is disabled during the call to torch_dispatch, this removes the `no_dispatch` calls that needed to be everywhere in order to prevent infinite recursion and will be necessary for nested modes
- nested modes are "allowed" (doesn't break) if the mode passed to the instance is the same (it's treated as a noop in this case). Otherwise, it will error like before
- Adds some `kwargs` that let a user override a current mode to a new one (not the same as nesting because it ignores the outer context manager)
- small string differences so the error messages are the same between the two

[ghstack-poisoned]

ezyang commented on torch/csrc/utils/python_mode.h:
```diff
@@ -0,0 +1,23 @@
+namespace torch {
+namespace python_dispatch {
+
+struct StashPythonModeGuard {
```
I guess this naming isn't symmetric with the torch function guard. Do you prefer Stash over No?

samdow replied:
I matched the name to [this guard in PythonFallback](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/PythonFallbackKernel.cpp#L30-L44) that has the same pattern. I guess I prefer it slightly but don't feel strongly. I can change torch function to be stash as well

ezyang commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -285,18 +286,30 @@ auto handle_torch_function_no_python_arg_parser(
-        TORCH_WARN("Defining your `__torch_function__` as a plain method is deprecated and ",
-                   "will be an error in future, please define it as a classmethod.");
+          torch_function.ptr() != torch::disabled_torch_function_impl() &&
+          !at::impl::PythonModeTLS::get_state()) {
```
If you handle mode separately in the block above, you won't need this condition anymore. This condition is a bit finely balanced right now in any case, because you are indirectly saying that this `torch_function` is coming from a mode if there exists a mode, but if the mode itself returned NotImplemented you could have ended up in subsequent loops where this invariant is no longer true...

ezyang commented on torch/overrides.py:
```diff
@@ -1726,7 +1714,7 @@ class TorchFunctionModeMeta(type):
     def __new__(metacls, name, bases, dct):
         if '__init__' in dct:
-            dct['__init__'] = _wrap_init(dct['__init__'])
+            dct['__init__'] = _wrap_init(dct['__init__'], "TorchFunctionMode", "torch_function")
```
Don't want to generalize this for both function and dispatch? ;)

ezyang replied:
oh this might just be dead code now?

samdow replied:
This ended up not actually being dead code. I definitely got confused on that too

> Don't want to generalize this for both function and dispatch? ;)

Sorry I think I'm parsing this weirdly--are you saying we shouldn't generalize the `_wrap_init`?

ezyang replied:
My main question is if you have a dispatch/function generic metaclass for modes, then why do we need TorchFunctionModeMeta

samdow replied:
Oh I see--ModeMeta is actually the dead code, not this (but also deleting that, I missed it)

I originally had that when I thought I could make this generic but that piece of code actually didn't work so I started passing everything as strings to avoid circular dependencies. At minimum, I'll put these strings into a little dataclass with some comments so it's clearer why they're being used (again mostly to avoid circular dependencies)

samdow replied:
Added the dataclass with a comment on the parent class but leaving this open in case we want more discussion

ezyang commented on torch/utils/_python_dispatch.py:
```diff
@@ -1,34 +1,156 @@
+
+
+@dataclass
+class PythonModeInfo(_ModeInfo):
```
yeah, my vote is for just fixing the naming all in one go in this PR

ezyang replied:
for ease in migration, you can keep `enable_python_mode` BC stub, but everything else should be in the new name. We should also decide what public module we're going to put this in.

ezyang commented on torch/overrides.py:
```diff
@@ -1790,6 +1778,29 @@ def _no_torch_function_mode() -> Iterator[None]:
+    def get_mode(self):
+        return _get_torch_function_mode()
+
+    def help_text(self, mode) -> str:
```
this is dead code now right

ezyang commented on torch/utils/_mode_utils.py:
```diff
@@ -0,0 +1,123 @@
+import functools
+from typing import Iterator
+from dataclasses import dataclass
+
```
would be nice to have a top level comment for the file

samdow replied:
Added!

ezyang commented on torch/utils/_mode_utils.py:
```diff
@@ -0,0 +1,123 @@
+        isinstance(mode, mode_info.mode_class) or
+        (isinstance(mode, type) and not issubclass(mode, mode_info.mode_class))
+    ):
+        raise ValueError(f'expected to get {mode_info.mode_class_name()}, Tensor-like class, '
```
for dispatch, the test for Tensor-like can be more stringent, as only subclasses of torch.Tensor can be modes. Still wondering if I should have just eliminated this case though lol

zou3519 replied:
Do you mean the case where we allow passing in subclasses of Tensor rather than a Mode object? 

If so, my vote is to eliminate those cases: the way someone writes a Mode's `__torch_function__/__torch_dispatch__` is different from the way someone writes a Tensor subclass's `__torch_function__/__torch_dispatch__`. In particular, for Tensors, (1) there's an implicit assumption that at least one of the Tensors being dispatched on is the Tensor subclass and (2) the torch_function/torch_dispatch on a Tensor subclass is recommended to be classmethod (so, it cannot store metadata unlike a Mode object)

samdow replied:
Stacked this on #76562, naming should be consistent now!

ezyang reviewed (approved):
A bunch of fiddly small stuff, but generally very happy with this PR. Thanks for doing it!

zou3519 commented on torch/csrc/utils/python_mode.h:
```diff
@@ -0,0 +1,23 @@
+
+struct StashPythonModeGuard {
+public:
+  StashPythonModeGuard() {
```
These are gonna need some renaming after https://github.com/pytorch/pytorch/pull/76562

zou3519 commented on torch/_C/__init__.pyi.in:
```diff
@@ -746,8 +746,15 @@ def __set_forward_AD_enabled(enabled: _bool) -> None: ...
+    ...
+
+def _set_torch_function_mode(cls: Optional[Union[type, TorchFunctionMode]]) -> None: ...
+def _get_torch_function_mode() -> Type: ...
```
Also probably should be `Optional[Union[type, TorchFunctionMode]`? (I'm not sure about the Optional part, what happens if there is no mode?)

ezyang replied:
it returns None

zou3519 reviewed (approved):
LGTM as well, modulo Ed's comments

Commit:
Update on "Enable nesting for python_mode"


This adds the ability to nest python_modes with `push_python_mode`, based on the torch_function design introduced [here](https://github.com/pytorch/pytorch/pull/75154). In order to push python modes, we needed to (1) introduce the concept of PythonMode (a la TorchFunctionMode) that defines a torch_dispatch but isn't a tensor subclass, (2) update enable_python_mode to match torch function mode (longer explanation below), and (3) introduce `push_python_mode` alongside the now updated `enable_python_mode`. Users will only be able to nest modes using push but not enable

**Longer note on what it means to update enable_python_mode**

`enable_python_mode`/`enable_torch_function_mode` are the simplest versions of the modes that don't support nesting but do support passing a tensor subclass as the mode. The current implementation of `enable_python_mode` didn't match the functionality of `enable_torch_function_mode` (which limited the ability to add nesting in the form of `push_python_mode` in the next PR)

This makes their behaviors match which means:
- the mode is disabled during the call to torch_dispatch, this removes the `no_dispatch` calls that needed to be everywhere in order to prevent infinite recursion and will be necessary for nested modes
- nested modes are "allowed" (doesn't break) if the mode passed to the instance is the same (it's treated as a noop in this case). Otherwise, it will error like before
- Adds some `kwargs` that let a user override a current mode to a new one (not the same as nesting because it ignores the outer context manager)
- small string differences so the error messages are the same between the two

[ghstack-poisoned]

samdow commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -285,18 +286,30 @@ auto handle_torch_function_no_python_arg_parser(
+            py_types.ptr(),
+            args,
+            kwargs,
+            NULL));
```
Ah I see! I didn't realize this was the intention and figured that something weird might be happening for torch function handling since it's using call method and some dummy strings, but totally worked to have it here. Reconfigured in the newest commit

samdow commented on test/test_python_dispatch.py:
```diff
@@ -497,9 +495,12 @@ def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
-        # B has precedence over A due to the subclass relationship
         with self.assertRaises(ErrorB):
+            a + b
+
+        # B has precedence over A due to the subclass relationship yet
+        # modes take precedence over arguments
+        with self.assertRaises(ErrorA):
```
Flagging that this behavior changed from the original version of torch dispatch (fka python) mode. IMO this behavior feels more like how I would expect modes to work, but just wanted to flag that it's a change in behavior

ezyang replied:
Thanks for the flag. This change looks fine. Even better if we just get rid of direct tensor subclass mode, that will eliminate the question here entirely.

Commit:
Update on "Enable nesting for python_mode"


This adds the ability to nest python_modes with `push_python_mode`, based on the torch_function design introduced [here](https://github.com/pytorch/pytorch/pull/75154). In order to push python modes, we needed to (1) introduce the concept of PythonMode (a la TorchFunctionMode) that defines a torch_dispatch but isn't a tensor subclass, (2) update enable_python_mode to match torch function mode (longer explanation below), and (3) introduce `push_python_mode` alongside the now updated `enable_python_mode`. Users will only be able to nest modes using push but not enable

**Longer note on what it means to update enable_python_mode**

`enable_python_mode`/`enable_torch_function_mode` are the simplest versions of the modes that don't support nesting but do support passing a tensor subclass as the mode. The current implementation of `enable_python_mode` didn't match the functionality of `enable_torch_function_mode` (which limited the ability to add nesting in the form of `push_python_mode` in the next PR)

This makes their behaviors match which means:
- the mode is disabled during the call to torch_dispatch, this removes the `no_dispatch` calls that needed to be everywhere in order to prevent infinite recursion and will be necessary for nested modes
- nested modes are "allowed" (doesn't break) if the mode passed to the instance is the same (it's treated as a noop in this case). Otherwise, it will error like before
- Adds some `kwargs` that let a user override a current mode to a new one (not the same as nesting because it ignores the outer context manager)
- small string differences so the error messages are the same between the two

[ghstack-poisoned]

Commit:
Update on "Enable nesting for python_mode"


This adds the ability to nest python_modes with `push_python_mode`, based on the torch_function design introduced [here](https://github.com/pytorch/pytorch/pull/75154). In order to push python modes, we needed to (1) introduce the concept of PythonMode (a la TorchFunctionMode) that defines a torch_dispatch but isn't a tensor subclass, (2) update enable_python_mode to match torch function mode (longer explanation below), and (3) introduce `push_python_mode` alongside the now updated `enable_python_mode`. Users will only be able to nest modes using push but not enable

**Longer note on what it means to update enable_python_mode**

`enable_python_mode`/`enable_torch_function_mode` are the simplest versions of the modes that don't support nesting but do support passing a tensor subclass as the mode. The current implementation of `enable_python_mode` didn't match the functionality of `enable_torch_function_mode` (which limited the ability to add nesting in the form of `push_python_mode` in the next PR)

This makes their behaviors match which means:
- the mode is disabled during the call to torch_dispatch, this removes the `no_dispatch` calls that needed to be everywhere in order to prevent infinite recursion and will be necessary for nested modes
- nested modes are "allowed" (doesn't break) if the mode passed to the instance is the same (it's treated as a noop in this case). Otherwise, it will error like before
- Adds some `kwargs` that let a user override a current mode to a new one (not the same as nesting because it ignores the outer context manager)
- small string differences so the error messages are the same between the two

[ghstack-poisoned]

samdow commented:
@pytorchbot merge this please
